src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/parquet/shredder.h"
+
+#include "serde/parquet/schema.h"
+#include "serde/parquet/value.h"
+#include "src/v/serde/parquet/schema.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <stdexcept>
+
+namespace serde::parquet {
+
+namespace {
+
+struct traversal_levels {
+    uint8_t repetition_depth;
```

Comment:
Done - does that help/make sense? It's basically the current number of repeated ancestors (self included) in the schema tree. That's needed because the actual repetition level for the first element needs to be different from this. We set the value of repetition_level to be this after the first element in the list.

Reply:
Thanks for the helpful comment! Been reading through the passage of the dremel paper about these levels and it's becoming much clearer for me